src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java

```diff
@@ -138,24 +142,77 @@ void setType(int parameterIndex, Integer type) throws SQLException {
+    setParameter(parameterIndex, value, null, null, null, null);
+  }
+
+  /**
+   * Sets a parameter value as the specified vendor-specific {@link SQLType}. Only {@link JsonType}
+   * is currently supported.
+   */
+  void setParameter(int parameterIndex, Object value, SQLType sqlType) throws SQLException {
+    setParameter(parameterIndex, value, null, null, null, sqlType);
+  }
+
+  /**
+   * Sets a parameter value as the specified vendor-specific {@link SQLType} with the specified
+   * scale or length. Only {@link JsonType} is currently supported, which does not have a variable
+   * scale or length. This method is only here to support the {@link
+   * PreparedStatement#setObject(int, Object, SQLType, int)} method.
+   */
+  void setParameter(int parameterIndex, Object value, SQLType sqlType, Integer scaleOrLength)
+      throws SQLException {
+    setParameter(parameterIndex, value, null, scaleOrLength, null, sqlType);
+  }
+
+  /**
+   * Sets a parameter value as the specified sql type. The type can be one of the constants in
+   * {@link Types} or a vendor specific type code supplied by a vendor specific {@link SQLType}.
+   * Currently only {@link JsonType#VENDOR_TYPE_NUMBER} is supported for the latter.
+   */
   void setParameter(int parameterIndex, Object value, Integer sqlType) throws SQLException {
     setParameter(parameterIndex, value, sqlType, null);
   }
 
+  /**
+   * Sets a parameter value as the specified sql type with the specified scale or length. The type
+   * can be one of the constants in {@link Types} or a vendor specific type code supplied by a
+   * vendor specific {@link SQLType}. Currently only {@link JsonType#VENDOR_TYPE_NUMBER} is
+   * supported for the latter.
+   */
   void setParameter(int parameterIndex, Object value, Integer sqlType, Integer scaleOrLength)
       throws SQLException {
-    setParameter(parameterIndex, value, sqlType, scaleOrLength, null);
+    setParameter(parameterIndex, value, sqlType, scaleOrLength, null, null);
   }
 
+  /**
+   * Sets a parameter value as the specified sql type with the specified scale or length. Any {@link
+   * SQLType} instance will take precedence over sqlType. The type can be one of the constants in
+   * {@link Types} or a vendor specific type code supplied by a vendor specific {@link SQLType}.
+   * Currently only {@link JsonType#VENDOR_TYPE_NUMBER} is supported for the latter.
+   */
   void setParameter(
-      int parameterIndex, Object value, Integer sqlType, Integer scaleOrLength, String column)
+      int parameterIndex,
+      Object value,
+      Integer sqlType,
+      Integer scaleOrLength,
+      String column,
+      SQLType sqlTypeObject)
       throws SQLException {
-    // check that only valid type/value combinations are entered
-    if (sqlType != null) {
-      checkTypeAndValueSupported(value, sqlType);
+    // Ignore the sql type if the application has created a Spanner Value object.
```

Comment:
That is an interesting design decision. So if we do `preparedStatement.setObject(Value.json("{...}"))` we would need to specify the `SqlType`. I like this.